--- newt-0.3.orig/newt.c	2004-12-29 10:33:04.125614200 +0100
+++ newt-0.3/newt.c	2004-12-29 10:41:10.317701768 +0100
@@ -298,7 +298,7 @@
 		zval **z_data_pnfk = NULL, res_pnfk; \
 		zend_hash_internal_pointer_reset (&NEWT_G(data)); \
 		while (zend_hash_get_current_data (&NEWT_G(data), (void **)&z_data_pnfk) == SUCCESS) { \
-			is_identical_function (&res_pnfk, (zval *) *z_data_pnfk, z_data); \
+			is_identical_function (&res_pnfk, (zval *) *z_data_pnfk, z_data TSRMLS_CC); \
 			if(zval_is_true (&res_pnfk)) { \
 				zend_hash_get_current_key(&NEWT_G(data), NULL, &key, 0); \
 				break; \
