diff -burN newt-0.2.orig/newt.c newt-0.2/newt.c
--- newt-0.2.orig/newt.c	2004-12-13 21:49:14.924280112 +0100
+++ newt-0.2/newt.c	2004-12-13 21:50:17.695737400 +0100
@@ -107,7 +107,9 @@
 	PHP_FE(newt_listbox_get_selection,			NULL)
 	PHP_FE(newt_listbox_clear_selection,		NULL)
 	PHP_FE(newt_listbox_select_item,			NULL)
+#ifdef HAVE_NEWT_LISTBOX_ITEM_COUNT
 	PHP_FE(newt_listbox_item_count,				NULL)
+#endif
 	PHP_FE(newt_checkbox_tree,					NULL)
 	PHP_FE(newt_checkbox_tree_multi,			NULL)
 	PHP_FE(newt_checkbox_tree_get_selection,	NULL)
diff -burN newt-0.2.orig/php_newt.h newt-0.2/php_newt.h
--- newt-0.2.orig/php_newt.h	2004-12-13 21:49:14.923280264 +0100
+++ newt-0.2/php_newt.h	2004-12-13 21:50:19.077527336 +0100
@@ -102,7 +102,9 @@
 PHP_FUNCTION(newt_listbox_get_selection);
 PHP_FUNCTION(newt_listbox_clear_selection);
 PHP_FUNCTION(newt_listbox_select_item);
+#ifdef HAVE_NEWT_LISTBOX_ITEM_COUNT
 PHP_FUNCTION(newt_listbox_item_count);
+#endif
 PHP_FUNCTION(newt_checkbox_tree);
 PHP_FUNCTION(newt_checkbox_tree_multi);
 PHP_FUNCTION(newt_checkbox_tree_get_selection);
